Features/473 implement merge operator

dmartensson opened the pull request:
Implements the merge operators and fixes 3 differences from less.js, issues 473, 521, 522 and 523.

Unit tests are probably not in the right scope and feel free to suggest where I should put them.

Commit:
Don´t leake information through AspResponseLogger.cs in production.

Commit:
Implemented '+' and '+_' merge operators

Commit:
[522] Error message "Expected 'x' found 'y'" displayed the wrong found char, it showed next char, not current

Commit:
[523] Empty variable declaration in root throws exception

This was caused by the Value method that consumed the ';' when looking if done, but containing method needed either ';' or '}' to consider the rule complete.

Commit:
Consume extra ';' after rules to follow Less.js behaviour

rytmis commented:
Awesome, thanks! I'll get back to you on this tomorrow.

rytmis commented on src/dotless.Core/Parser/Tree/Ruleset.cs:
```diff
@@ -279,6 +281,50 @@ protected void EvaluateRules(Env env)
+                }
+                else
+                {
+                    if (rule.Merge == "")
```
Is this a possible branch? Looks like the check on line 295 already covers this case.

dmartensson replied:
Yes, this was redundant.

rytmis commented on src/dotless.Core/Parser/Tree/Ruleset.cs:
```diff
@@ -279,6 +281,50 @@ protected void EvaluateRules(Env env)
+            foreach (var node in Rules)
+            {
+                var rule = node as Rule;
+                if (rule == null || rule.Merge == "")
```
Performance implications aside, you could pull out the rules that match this condition by doing a

```csharp
var rulesToMerge = Rules.OfType<Rule>().Where(r => !string.IsNullOrEmpty(r.Merge));
var sorted = Rules.Except(rulesToMerge).Cast<Node>().ToList()
```

prior to the loop.

dmartensson replied:
I will look at it, but if a later rule overwrites the merged property ordering is important so I did not wish to break it apart.

For a static stylesheet you could get away with it but if the stylesheet is in multiple parts that are joined before parsed you might have overlapping rules both setting the same property and I elected to play it safe in this case.

But I will experiment with some more unit tests to see if I can design one that exemplifies this and then experiment with it and see if I can solve it better.

rytmis replied:
Ah, right, this would affect ordering as well. I'm still fairly sure it's possible to preserve ordering while still going for a more functional style. Since it's the ordering that currently drives this particular organization for the code, could you perhaps also add a test case for verifying the correct ordering?

dmartensson replied:
I have managed to remove the second loop and have that integrated with the first and are looking into simplifying it a bit more.

rytmis commented on src/dotless.Core/Parser/Tree/Ruleset.cs:
```diff
@@ -279,6 +281,50 @@ protected void EvaluateRules(Env env)
+                if (!groups.ContainsKey(key))
+                {
+                    groups[key] = new List<Rule> {rule};
+                    sorted.Add(node);
```
As a general rule, I'd much prefer an approach where this single loop would only concern itself with populating the `groups` dictionary. Is this interleaved here to preserve rule ordering or is it incidental?

dmartensson replied:
Order needs to be preserved since multiple rules for the same property must me merged or overwritten in the correct order for the selector.

Ex.

`
@paddingTopBottom:10px;
@paddingRightLeft:20px;
.padding(num)  { padding+_: num; }
div {
    .padding(@paddingTopBottom);
    .padding(@paddingRightLeft);
}`
(In this case the two padding rules might also come through more convoluted paths so it might be as obvious as this.)
 
This, after expands, ends up:
`
div {
    .padding+_:10px;
    .padding+_20px;
}`

And is merged to
`
    div { padding: 10px 20px; }`

If they get out of order the padding will be switched.

But I can look at the code and see if I can make it simpler or at least explain what is done better.

The first element in each group are added to the sorted list directly and when I then merge the group the element is already in the correct position in the resulting set of rules.
Any following merge rules for the same selector is only added to the merge group, not to the sorted output.

rytmis commented on src/dotless.Core/Parser/Tree/Rule.cs:
```diff
@@ -14,6 +16,12 @@ public class Rule : Node
+
+        public bool Important
+        {
+            get { return Value is Value && ((Value) Value).Important == "!important"; }
```
I'd go with `(Value as Value)?.Important == "!important"`

dmartensson replied:
Will do, this is the first place I got a bit uncertain if I could use c#6 features or not ;)

rytmis commented on src/dotless.Core/Parser/Parsers.cs:
```diff
@@ -93,7 +93,15 @@ public NodeList Primary(Parser parser)
+                        rule.Merge = rule.Name.EndsWith("+") ? ", " : " ";
+                        rule.Name = rule.Name.TrimEnd('+', '_');
+                    }
+                    root.Add(node); //TODO lägg till stöd för merge noder
```
Is this comment still needed? If it is, I'd translate it to English. ;-)

dmartensson replied:
No, your right, I just forgot to remove it.

rytmis commented on src/dotless.Test/Specs/RulesetsFixture.cs:
```diff
@@ -135,5 +135,135 @@ public void IESlash9HackAllowedForNonEvaluatedValues()
             var expected = input;
             AssertLess(input, expected);
         }
+
```
Having the tests in RulesetsFixture is fine by me -- what I'd like to see is a test or two covering cases where !important is involved.

dmartensson replied:
Your quite right, I should have made more cases to cover that, did not need it for the problem I had to solve but I imagine we will hit that to very soon now that we have solved the original problem.

dmartensson replied:
Checked the less.js handling of !important and it is currently broken and there is a bug logged for it.
http://less2css.org/#%7B%22less%22%3A%22.padding()%20%7B%20%5Cn%20%20padding%2B_%3A%2010px%20!important%3B%20%5Cn%7D%5Cn%5Cndiv%20%7B%5Cn%20%20.padding()%3B%5Cn%20%20.padding()%3B%5Cn%7D%22%7D

https://github.com/less/less.js/issues/2367

But checking with w3c validator there should be a single !important in the end so I will go with that.

Also, I found that I made a mistake, + and +_ should not be separate groups according to hos the less.js version does it so I will change this.

dmartensson replied:
Question. Currently its AppendMany in output that actually does the concatenation of values, but to be able to use separate chars for joining the values I would either have to move this up into the AppendCSS method, or make the output.AppendMany look at each item to be appended for the relevant join char?

Example

a {
font+: italic;
font+: 1em;
font+: Arial;
font+_: Verdana;
}

This should result in
font: italic 1em Arial, Verdana;

Notice the , between Arial and Verdana.

I am inclined to work with output to not break overall structure but do you see any problem with adding that kind of awareness to the output class method?

rytmis replied:
Adding that logic to Output seems fine to me (perhaps by introducing an overload for AppendMany, since there seem to be quite a few of them already). And even if that turns out to be sub-optimal, that's far from the worst offender. This codebase has some gnarly bits in it. :)

rytmis commented on src/dotless.Core/Parser/Tree/Ruleset.cs:
```diff
@@ -279,6 +281,50 @@ protected void EvaluateRules(Env env)
+                var rule = @group[0];
+                var value = rule.Value as Value;
+                if (value == null)
+                    value = (Value) (rule.Value = new Value(new[] {rule.Value}, "", rule.Merge));
```
These couple of lines here really need some clarification while you still remember what's going on. It's not immediately obvious that this seems to cover a case where Rule.Value is not a Value. Also, while I love terse code, this line in particular has way too many meanings for the word Value to be readable, so maybe expand it a little bit? :)

dmartensson replied:
The case is when value is a number and not a value node.

Numbers contain only one value but the value object holds a list of values, for example border can have 3 values, width, style and color.

But I will refactor to better indicate what is going on.

The value object already handles the merging of the different separate values so I just rebuild the separate rules into one rule with multiple values and let the existing value class do the actual concatenation instead of creating a new almost identical class.

But  padding+_: 10px; will create a rule with a single numeric value that cannot take any extra so I have to replace the num value with a values class value containing the original num value as first element.

rytmis commented:
All in all, many thanks for your effort -- this is not the easiest codebase in the world to work with. :)

Commit:
Removed left over todo comment

rytmis commented:
Dumb question maybe, but are you still waiting for my comments on something before proceeding further? If so, I must have missed that.

dmartensson commented:
No,

Just was a bit hectic in the end of the week and a fully planned weekend.

Been looking into exactly how to get output to do what I need without to much of an ugly hack, we do not want any random glitches due to brittle implementation ;)

rytmis commented:
Yeah, no worries. Take your time. It's just that I've been so mired in my own work recently that I thought it entirely possible I've missed something, and wanted to check in.

dmartensson commented:
Seems like the Evaluate methods manages to unwrap the Value node I use to wrap the number to keep the merge character :/.

Some where in the closure evaluation but I have yet to find the exact point.

This means that I never even have the info when I start to actually merge the values.

So it will take a little time to go through it in the debugger but I am narrowing it down.

rytmis commented:
Yeah, that kind of trawling and debugging sound _very_ familiar. Many thanks for the time you have put into this so far!

twenzel committed:
Use editorconfig from roslyn team

twenzel committed:
Merge pull request #539 from dotless/develop

Update appveyor.yml

twenzel committed:
Merge pull request #540 from dotless/develop

Update appveyor settings

twenzel committed:
Fixed artifact handling

twenzel committed:
Update the nuget api key

twenzel committed:
Rename the AspNet package

twenzel committed:
Update readme.md

Add package description to readme

twenzel committed:
Fixes nuspec

twenzel committed:
Update test packages

twenzel committed:
Update version number handling

twenzel committed:
Updated assembly info

twenzel committed:
Set version number in assemblies

twenzel committed:
Update readme.md

Add framework support description

twenzel committed:
Updated sample config

twenzel committed:
Merge branch 'master' of https://github.com/dotless/dotless

twenzel committed:
Fix wrong assembly name in web.config transformation #548
Insert section at correct place #549

twenzel committed:
Add strong name for dotless.AspNet #552

twenzel committed:
Fix transformation

twenzel committed:
Initial version

twenzel committed:
Add support for 4.5, 4.6.1 and 4.7
Also uses the System.Drawing #551

twenzel committed:
Update Readme

Wolf-Code committed:
Color hex usage, and Quoted not being cloned properly (#556)

* Fixed color(hex) throwing exception.

* Fixed Quoted not being cloned properly.

* Undid formatting changes.

* Undid formatting changes.

* Review changes:
- Added spec for mixin with content and not using parantheses.
- Renamed color hex test
- Fixed color test file namespace

twenzel committed:
Merge pull request #554 from dotless/feature/EnableNET45

Enable net45

twenzel commented:
@rytmis Should we merge this or do we still waiting for @dmartensson ?

twenzel committed:
Fix tests

twenzel committed:
Update NuGet packages, remove unused packages

twenzel committed:
Fix nuspecs

twenzel committed:
Merge pull request #557 from dotless/features/UpdateDependencies

Update NuGet packages, remove unused packages

twenzel committed:
Fix build version issue

rytmis commented:
@twenzel As I understood @dmartensson's last comment, it wasn't quite working yet, so I wouldn't quite pull it in yet.

twenzel committed:
- Fixed importer to get benchmark working
- Fixed Benchmark project

twenzel committed:
Fix importer error #555

twenzel committed:
Fix importer error #555

twenzel committed:
Fix importer error #555

twenzel committed:
Fix column alignment

twenzel committed:
Merge branch 'fix/EmptyResponse' of https://github.com/dotless/dotless into fix/EmptyResponse

Tigraine committed:
Merge pull request #538 from twenzel/feature/NewEditorConfig

Use editorconfig from roslyn team

twenzel committed:
Merge pull request #558 from dotless/feature/GetBenchmarkWorking

Get benchmark project working again

twenzel committed:
Merge branch 'fix/EmptyResponse' of https://github.com/dotless/dotless into fix/EmptyResponse

twenzel committed:
- Register Parser and Importer as Transient
- Fix typo

twenzel committed:
Merge pull request #559 from dotless/fix/EmptyResponse

Fix empty response

dmartensson commented:
Yes, merge works for either space or , but you cannot use them together which is needed for font which can have both separators.

I failed to find a way to get this to work.

My current implementation uses a wrapper that keeps the separator char but any attempt to nest merge was lost due to rewriting of rules.
And a single wrapper cannot handle multiple separators in this way since I also found no way to group children for each separator.

Currently I have no time to work on this at work unfortunately since we do not need both in our situation.

rocklan committed:
Added support for being able to supply an instance of a logger

dmartensson committed:
Merge pull request #1 from dotless/master

Updating my fork with your changes

twenzel commented:
@dmartensson The tests are located in dotless/tests/dotless.Core.Test/Specs/

twenzel committed:
Updated Nuget key

twenzel committed:
Fix typo in description

dmartensson committed:
Merge pull request #2 from dotless/master

Pull changes from dotless

Tigraine committed:
Merge pull request #285 from guidmaster/master

Don´t leake information through AspResponseLogger.cs in production.

Tigraine committed:
Merge pull request #566 from rocklan/master

Added support for being able to supply an instance of a logger

snipervld committed:
importer: reset reference imports in ResetImports method

snipervld committed:
tests: add test for bug in ResetImports

twenzel committed:
Merge pull request #594 from snipervld/fix/reset-reference-imports

Fixes the issue with ResetImports method and reference imports

twenzel committed:
Create testbuild.yml

twenzel committed:
Update testbuild.yml

twenzel committed:
Update build.cake

twenzel committed:
Update build.cake

twenzel committed:
Update build.cake

twenzel committed:
Fix msbuild locator

twenzel committed:
Update testbuild.yml

twenzel committed:
Merge pull request #595 from dotless/add-pr-build

Create testbuild.yml

andersjonsson committed:
Merge branch 'dotless:master' into master

Commit:
merge master

Commit:
fixed test
